### DISCUSSION TITLE
Cleanup of refs/branchless/xxx refs

### DISCUSSION THREAD

**Discussion by talios** [arxanas/git-branchless#1268] [General]
Should we expect git-branchless to automatically cleanup lingering refs/branchless/xxx entries over time at all?

**Comment by arxanas** [comment]
git-branchless installs a [`pre-auto-gc`](https://git-scm.com/docs/githooks#_pre_auto_gc) hook that cleans up the refs under `refs/branchless/`. This hook triggers before automatic GC, or you can run `git branchless gc` manually as well.

However, this hook/command only cleans up references which point to abandoned commits. By itself, the number of references that git-branchless installs is more than necessary just to keep the commits live. One alternative approach to "compact" the references would be to only create/preserve references to the "heads" of the live commits (those which have no descendants), which would be sufficient to keep everything live, but also substantially reduce the number of live commits.

The current logic is fairly straightforward, and I think the above change would not be difficult to implement, if you're interested.

https://github.com/arxanas/git-branchless/blob/3d8b480ad8006079eb97d73d9ef76a81924fd545/git-branchless-lib/src/core/gc.rs#L23-L128
